pkg/csi/service/vanilla/controller.go

```diff
@@ -446,33 +446,45 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
 
 	resp := &csi.CreateVolumeResponse{
 		Volume: &csi.Volume{
-			VolumeId:      volumeID,
+			VolumeId:      volumeInfo.VolumeID.Id,
 			CapacityBytes: int64(units.FileSize(volSizeMB * common.MbInBytes)),
 			VolumeContext: attributes,
 		},
 	}
 
-	// Call QueryVolume API and get the datastoreURL of the Provisioned Volume
+	// Retrieve the datastoreURL of the Provisioned Volume
+	// If CNS CreateVolume API does not return datastoreURL, retrieve this by calling QueryVolume
+	// otherwise, retrieve this from PlacementResults from the response of CreateVolume API
 	var volumeAccessibleTopology = make(map[string]string)
+	var datastoreAccessibleTopology = make([]map[string]string, 0)
+	var datastoreURL string
 	if len(datastoreTopologyMap) > 0 {
-		volumeIds := []cnstypes.CnsVolumeId{{Id: volumeID}}
-		queryFilter := cnstypes.CnsQueryFilter{
-			VolumeIds: volumeIds,
-		}
-		queryResult, err := c.manager.VolumeManager.QueryVolume(ctx, queryFilter)
-		if err != nil {
-			log.Errorf("QueryVolume failed for volumeID: %s", volumeID)
-			return nil, status.Error(codes.Internal, err.Error())
-		}
-		if len(queryResult.Volumes) > 0 {
-			// Find datastore topology from the retrieved datastoreURL
-			datastoreAccessibleTopology := datastoreTopologyMap[queryResult.Volumes[0].DatastoreUrl]
-			log.Debugf("Volume: %s is provisioned on the datastore: %s ", volumeID, queryResult.Volumes[0].DatastoreUrl)
-			if len(datastoreAccessibleTopology) > 0 {
-				rand.Seed(time.Now().Unix())
-				volumeAccessibleTopology = datastoreAccessibleTopology[rand.Intn(len(datastoreAccessibleTopology))]
-				log.Debugf("volumeAccessibleTopology: [%+v] is selected for datastore: %s ", volumeAccessibleTopology, queryResult.Volumes[0].DatastoreUrl)
+		if volumeInfo.DatastoreURL == "" {
+			volumeIds := []cnstypes.CnsVolumeId{{Id: volumeInfo.VolumeID.Id}}
+			queryFilter := cnstypes.CnsQueryFilter{
+				VolumeIds: volumeIds,
+			}
+			queryResult, err := c.manager.VolumeManager.QueryVolume(ctx, queryFilter)
+			if err != nil {
+				log.Errorf("QueryVolume failed for volumeID: %s", volumeInfo.VolumeID.Id)
+				return nil, status.Error(codes.Internal, err.Error())
 			}
+			if len(queryResult.Volumes) > 0 {
+				// Find datastore topology from the retrieved datastoreURL
+				datastoreAccessibleTopology = datastoreTopologyMap[queryResult.Volumes[0].DatastoreUrl]
+				datastoreURL = queryResult.Volumes[0].DatastoreUrl
+				log.Debugf("Volume: %s is provisioned on the datastore: %s ", volumeInfo.VolumeID.Id, datastoreURL)
+			}
+		} else {
+			// retrieve datastoreURL from placementResults
+			datastoreAccessibleTopology = datastoreTopologyMap[volumeInfo.DatastoreURL]
+			datastoreURL = volumeInfo.DatastoreURL
+			log.Debugf("Volume: %s is provisioned on the datastore: %s ", volumeInfo.VolumeID.Id, datastoreURL)
+		}
+		if len(datastoreAccessibleTopology) > 0 {
```

Comment:
Could you explain why we are selecting a random entry from `datastoreAccessibleTopology`?

Reply:
That is the existing logic. I think @divyenpatel know more about it.